File: ctci/chapter_1_arrays_strings/js/1.1-isUnique.js
/* O(n) solution */

/**
 * Determines if an ASCII string has all unique characters.
 * @param   {string}  str
 * @returns {boolean}
 */
const isUnique = str => {
  const map = {}
  for (let char of str) {
    if (map[char]) return false
    map[char] = true
  }
  return true
}

console.log(isUnique("abcdef")) // true
console.log(isUnique("abcdea")) // false
console.log(isUnique("aa")) // false
console.log(isUnique("a")) // true
console.log(isUnique("abc123")) // true
console.log(isUnique("")) // true

File: ctci/chapter_1_arrays_strings/1.1-is_unique.py
# O(n) solution


def is_unique(string: str) -> bool:
    """
    Determines if an ASCII string has all unique characters.
    """
    _dict = {}
    for char in string:
        if char in _dict:
            return False
        _dict[char] = True
    return True


if __name__ == '__main__':
    print(is_unique("abcdef"))  # true
    print(is_unique("abcdea"))  # false
    print(is_unique("aa"))  # false
    print(is_unique("a"))  # true
    print(is_unique("abc123"))  # true
    print(is_unique(""))  # true
